src/peergos/server/storage/S3BlockStorage.java

```diff
@@ -0,0 +1,335 @@
+        return Futures.of(Collections.emptyList());
+    }
+
+    @Override
+    public CompletableFuture<List<Multihash>> recursiveUnpin(PublicKeyHash owner, Multihash hash) {
+        return Futures.of(Collections.emptyList());
+    }
+
+    @Override
+    public CompletableFuture<Boolean> gc() {
+        return Futures.errored(new IllegalStateException("S3 doesn't implement GC!"));
+    }
+
+    @Override
+    public CompletableFuture<Optional<Integer>> getSize(Multihash hash) {
+        return Futures.of(map(hash, h -> {
+            GetObjectRequest get = new GetObjectRequest(bucket, folder + hashToKey(hash));
+            Histogram.Timer readTimer = readTimerLog.labels(h.toString()).startTimer();
+            try (S3Object res = s3Client.getObject(get)) {
+                return Optional.of((int)res.getObjectMetadata().getContentLength());
+            } catch (IOException e) {
+                throw new RuntimeException(e.getMessage(), e);
+            } finally {
+                readTimer.observeDuration();
+            }
+        }, e -> Optional.empty()));
+    }
+
+    public boolean contains(Multihash key) {
+        return map(key, h -> s3Client.doesObjectExist(bucket, folder + hashToKey(h)), e -> false);
+    }
+
+    public <T> T map(Multihash hash, Function<Multihash, T> success, Function<Throwable, T> absent) {
+        try {
+            return success.apply(hash);
+        } catch (AmazonServiceException e) {
+            /* Caught an AmazonServiceException,
+               which means our request made it
+               to Amazon S3, but was rejected with an error response
+               for some reason.
+            */
+            if ("NoSuchKey".equals(e.getErrorCode())) {
+                Histogram.Timer readTimer = readTimerLog.labels("0").startTimer();
+                readTimer.observeDuration();
+                return absent.apply(e);
+            }
+            LOG.warning("AmazonServiceException: " + e.getMessage());
+            LOG.warning("AWS Error Code:   " + e.getErrorCode());
+            throw new RuntimeException(e.getMessage(), e);
+        } catch (AmazonClientException e) {
+            /* Caught an AmazonClientException,
+               which means the client encountered
+               an internal error while trying to communicate
+               with S3, such as not being able to access the network.
+            */
+            LOG.severe("AmazonClientException: " + e.getMessage());
+            LOG.severe("Thrown at:" + e.getCause().toString());
+            throw new RuntimeException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Multihash> id() {
+        return Futures.of(id);
+    }
+
+    @Override
+    public CompletableFuture<TransactionId> startTransaction(PublicKeyHash owner) {
+        // TODO delegate this to a sql table
+        return CompletableFuture.completedFuture(new TransactionId(Long.toString(System.currentTimeMillis())));
```

Comment:
Risk of duplicate transactionId. Resolution of currentTimeMillis is very low.

Reply:
That's okay until we have a sql version because they don't do anything.